design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,655 @@
+        # please note, there is no namespace option here, platform consumers
+        # only have access to their team namespaces, so the StoreConfig in the
+        # same namespace as Claim will be used.
+```
+
+
+**Claim** spec for writing connection secret to both Kubernetes and Vault:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: database-creds
+    # this is to indicate that we want to store a kubernetes
+    # secret, but we have no config to set, will write into
+    # the same namespace.
+    kubernetes: {} 
+    externalStore:
+      configRef:
+        name: vault-team-a
+```
+
+#### External Secret Store Configuration: StoreConfig
+
+External secret store configuration will contain the required information other
+than the name (and optional attributes) of secret. Thanks to
+[the standardization efforts] on a declarative API for syncing secrets from
+external stores into Kubernetes, there is already an [existing schema] that we
+can follow here. However, there will be slight differences since we want to
+receive the `name` of secret as input in our API whereas, they expect a full
+identifier of the secret instance within the secret store
+(in [ExternalSecret spec] at `spec.dataFrom.key`). Another reason for
+differences is the direction of operation; Crossplane needs to _publish to_
+external stores whereas those tools targets the opposite, that is
+_fetching from_ external stores.
+
+**Examples:**
+
+Publish to Vault under parent path `secret/my-cloud/dev/` using Kubernetes auth:
+
+```yaml
+apiVersion: secrets.crossplane.io/v1alpha1
+kind: StoreConfig
+metadata:
+  name: vault-platform
+  namespace: crossplane-system
+spec:
+  type: Vault
+  vault:
+    server: "https://vault.acme.org"
+    # parentPath is the parent path that will be prepended to the secrets
+    # created with this store config.
+    parentPath: "secret/my-cloud/dev/"
+    version: "v2"
+    caBundle: "..."
+    
+    auth:
+      # Kubernetes auth: https://www.vaultproject.io/docs/auth/kubernetes
+      kubernetes:
+        mountPath: "kubernetes"
+        role: "demo"
+```
+
+Publish with an _out-of-tree Secret Plugin_ (for future support, if needed):
+
+```yaml
+apiVersion: secrets.crossplane.io/v1alpha1
+kind: StoreConfig
+metadata:
+  name: vault-platform
+  namespace: crossplane-system
+spec:
+  type: Plugin
+  plugin:
+    name: plugin-x
+    endpoint: unix:///tmp/plugin-x.sock
+    config:
+      host: secretstore.acme.org
+      port: 9999
+      caBundle: "..."
+      some: 
+        other:
+          arbitrary-config: true
+```
+
+### User Experience
+
+With the API definition above, it is mostly clear how the user interaction would
+be. But one important point that worth mentioning is, both Crossplane core pod
+and provider pods would need to access to the external secret stores. The
+credentials that is made available needs to be authorized to read, write and
+delete the secrets living at the configured scope. 
```

Comment:
I am not happy with introducing the only dependency between provider -> core Crossplane, but, UX-wise, I don't feel comfortable regarding the duplication. I wouldn't be too concerned regarding Provider or Crossplane not supporting a new store (yet) as long as we return a meaningful error message that users could easily find (i.e. via events). 
Putting implementation details aside, this sounds like a global configuration rather than a provider-specific one (different than `ProviderConfig`).
Also, I am wondering if we could just use the unstructured client to get `StoreConfig` in the providers not to have an explicit dependency and only look for it if a StoreConfig referenced (otherwise store as a k8s secret with some defaulting as you suggested). Or somehow put this type to crossplane-runtime so that we can register its schema from providers but install its CRD as part of Core Crossplane.

In any case, we can leverage being an alpha feature and revisit the decision we made here based on the community feedback after some baking period.